More sharded_tensor creation ops: harded_tensor.zeros, sharded_tensor.full, sharded_tensor.rand

bowangbj opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #63732

Summary:

Test Plan:

$ python test/distributed/_sharded_tensor/test_sharded_tensor.py  --v

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D30472621](https://our.internmc.facebook.com/intern/diff/D30472621)

cc @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @agolynski @SciPioneer @H-Huang @mrzzd @cbalioglu @gcramer23

bowangbj committed:
More sharded_tensor creation ops: harded_tensor.zeros, sharded_tensor.full, sharded_tensor.rand

Summary:

Test Plan:

$ python test/distributed/_sharded_tensor/test_sharded_tensor.py TestCreateTensorFromParams --v
$ python test/distributed/_sharded_tensor/test_sharded_tensor.py TestShardedTensorChunked --v

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

bowangbj commented:
@bowangbj has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30472621).

bowangbj committed:
lint on "More sharded_tensor creation ops: harded_tensor.zeros, sharded_tensor.full, sharded_tensor.rand"


Summary:

Test Plan:

$ python test/distributed/_sharded_tensor/test_sharded_tensor.py  --v

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D30472621](https://our.internmc.facebook.com/intern/diff/D30472621)

[ghstack-poisoned]

bowangbj commented:
@bowangbj has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30472621).

pritamdamania87 commented on torch/distributed/_sharded_tensor/api.py:
```diff
@@ -127,17 +127,22 @@ def _register_remote_shards(sharded_tensor_id: int, rrefs: List[rpc.RRef[Shard]]
 
     create_op: CreateOp
+    # needed when create_op is FULL
+    fill_value: Number
```
Since this is optional, can we specify a default value: `fill_value: Number = None`, this way users don't need to pass this in unless they are using `CreateOp::FULL`.

bowangbj replied:
last time I got some flake8 complaint, but yes it made sense to set default as None

pritamdamania87 commented on torch/distributed/_sharded_tensor/api.py:
```diff
@@ -667,21 +672,43 @@ def __setstate__(self, state):
+                          device=local_device, )
     else:
         raise ValueError(f'Unsupported create_op: {tensor_init_params.create_op}')
+        torch.rand
```
nit: is this a typo?

bowangbj replied:
Great catch

pritamdamania87 commented on torch/distributed/_sharded_tensor/api.py:
```diff
@@ -127,17 +127,22 @@ def _register_remote_shards(sharded_tensor_id: int, rrefs: List[rpc.RRef[Shard]]
     create_op: CreateOp
+    # needed when create_op is FULL
+    fill_value: Number
     dtype: torch.dtype
```
Actually it might be useful to have default values for all of these parameters exception  `create_op` so it is simpler to create TensorInitParams.

bowangbj replied:
SG, done

pritamdamania87 commented on torch/distributed/_sharded_tensor/__init__.py:
```diff
@@ -12,16 +12,15 @@
-        memory_format=torch.contiguous_format,
-        process_group=None,
-        init_rrefs=False):
+def empty(sharding_spec: ShardingSpec,
```
nit: It is easier to review code if formatting changes are in a separate PR.

bowangbj replied:
Since empty is the only one which does not align with the format of rest. Maybe just keep it since it's a tiny change?

pritamdamania87 commented on torch/distributed/_sharded_tensor/__init__.py:
```diff
@@ -52,7 +51,8 @@ def empty(
         A :class:`ShardedTensor` object on each rank
     """
-    tensor_init_params = TensorInitParams(create_op=CreateOp.EMPTY, dtype=dtype, layout=layout,
+    tensor_init_params = TensorInitParams(create_op=CreateOp.EMPTY, fill_value=0,
```
Do we need to specify `fill_value` here? We can omit it and it will just be `None` by default?

bowangbj replied:
Great point, done

pritamdamania87 commented on torch/distributed/_sharded_tensor/__init__.py:
```diff
@@ -101,7 +100,58 @@ def ones(
         A :class:`ShardedTensor` object on each rank
     """
-    tensor_init_params = TensorInitParams(create_op=CreateOp.ONES, dtype=dtype, layout=layout,
+    tensor_init_params = TensorInitParams(create_op=CreateOp.ONES, fill_value=0,
```
Same comment as above, can we omit fill_value? Also we are creating ones here so fill_value=0 doesn't seem right.

bowangbj replied:
Great point, removed.

pritamdamania87 commented on torch/distributed/_sharded_tensor/__init__.py:
```diff
@@ -101,7 +100,58 @@ def ones(
+         process_group=None,
+         init_rrefs=False):
+    """
+    Creates a rand :class:`ShardedTensor`. Needs to be called on all ranks in an SPMD fashion.
```
I think we can copy some of the docs from `torch.rand()` here (see https://pytorch.org/docs/stable/generated/torch.rand.html):

```
Returns a tensor filled with random numbers from a uniform distribution on the interval [0, 1)[0,1)
```

bowangbj replied:
Good point, done.

pritamdamania87 commented on torch/distributed/_sharded_tensor/__init__.py:
```diff
@@ -112,6 +162,109 @@ def ones(
+          process_group=None,
+          init_rrefs=False):
+    """
+    Creates a zeros :class:`ShardedTensor`. Needs to be called on all ranks in an SPMD fashion.
```
Can use the docs from `torch.zeros`: `Returns a tensor filled with the scalar value 0`. For ShardedTensor it could be something like:

```
Returns a :class:`ShardedTensor` filled with the scalar value 0. Needs to be called on all ranks in an SPMD fashion.
```

bowangbj replied:
Done

pritamdamania87 commented on torch/distributed/_sharded_tensor/__init__.py:
```diff
@@ -112,6 +162,109 @@ def ones(
+         process_group=None,
+         init_rrefs=False):
+    """
+    Creates a full :class:`ShardedTensor`. Needs to be called on all ranks in an SPMD fashion.
```
Use `torch.full` docs: `Creates a tensor of size size filled with fill_value. The tensor’s dtype is inferred from fill_value.`

It looks like torch.full allows you to specify a dtype but it still mentions this: "The tensor’s dtype is inferred from fill_value.". We should figure out what the semantics are for torch.full (ex: which dtype is used if fill_value dtype is different from dtype passed in) and mimic those here.

bowangbj replied:
if dtype is specified in torch.full(..), the type will be overridden. Otherwise, it will be inferred from fill_value. Our behavior is the same.  TestCreateTensorFromParams.test_full illustrates it.

bowangbj committed:
resolve pritam comment on "More sharded_tensor creation ops: harded_tensor.zeros, sharded_tensor.full, sharded_tensor.rand"


Summary:

Test Plan:

$ python test/distributed/_sharded_tensor/test_sharded_tensor.py  --v

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D30472621](https://our.internmc.facebook.com/intern/diff/D30472621)

[ghstack-poisoned]

bowangbj committed:
lint on "More sharded_tensor creation ops: harded_tensor.zeros, sharded_tensor.full, sharded_tensor.rand"


Summary:

Test Plan:

$ python test/distributed/_sharded_tensor/test_sharded_tensor.py  --v

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D30472621](https://our.internmc.facebook.com/intern/diff/D30472621)

[ghstack-poisoned]

bowangbj commented:
@bowangbj has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30472621).

bowangbj commented:
Thanks Pritam for a great, detailed and fast review. Rebased and fixed all your cmts. PTAL

pritamdamania87 commented on test/distributed/_sharded_tensor/test_sharded_tensor.py:
```diff
@@ -207,21 +203,27 @@ def test_rand(self):
             h, w, local_device=local_device, tensor_init_params=tensor_init_params)
-        expected_tensor = torch.full((h, w), fill_value=fill_value, device=local_device, dtype=expected_dtype)
+
+        print(f"=== bowangbj === \n local_tensor: {local_tensor}")
```
nit: can remove this?

bowangbj replied:
for some reason, seems you might look at the pre-lint version :) 
removed for real

pritamdamania87 commented on test/distributed/_sharded_tensor/test_sharded_tensor.py:
```diff
@@ -207,21 +203,27 @@ def test_rand(self):
+        fill_value = 23.5
+        tensor_properties = TensorProperties(
+            # tensor's dtype can be inferred from fill_value
+            dtype=None,
```
Shouldn't we have also test where we override the dtype here?

bowangbj replied:
Great point, added self.assertEqual(torch.float32, local_tensor.dtype) to test it explicitly

pritamdamania87 commented on torch/distributed/_sharded_tensor/__init__.py:
```diff
@@ -124,7 +126,8 @@ def rand(sharding_spec: ShardingSpec,
     """
-    Creates a rand :class:`ShardedTensor`. Needs to be called on all ranks in an SPMD fashion.
+    Returns a :class:`ShardedTensor` filled with random numbers from a uniform distribution on the
+        interval [0, 1)[0,1). Needs to be called on all ranks in an SPMD fashion.
```
nit: You actually want to use the following docstring here: ":math:`[0, 1)`". See https://github.com/pytorch/pytorch/blob/master/torch/_torch_docs.py#L7756.

bowangbj replied:
great ptr.

pritamdamania87 commented on torch/distributed/_sharded_tensor/__init__.py:
```diff
@@ -225,7 +231,8 @@ def full(sharding_spec: ShardingSpec,
     """
-    Creates a full :class:`ShardedTensor`. Needs to be called on all ranks in an SPMD fashion.
+    Creates a :class:`ShardedTensor` filled with fill_value. The tensor’s dtype
+        is inferred from fill_value. Needs to be called on all ranks in an SPMD fashion.
```
We should clarify the behavior clearly here that if dtype is specified that overrides the type inferred from fill_value.

bowangbj replied:
updated

pritamdamania87 reviewed (commented):
Looks good, just had a few more comments and it looks like there are some lint failures.

bowangbj committed:
resolve pritam comment on "More sharded_tensor creation ops: harded_tensor.zeros, sharded_tensor.full, sharded_tensor.rand"


Summary:

Test Plan:

$ python test/distributed/_sharded_tensor/test_sharded_tensor.py  --v

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D30472621](https://our.internmc.facebook.com/intern/diff/D30472621)

[ghstack-poisoned]

bowangbj commented:
@bowangbj has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30472621).

bowangbj commented:
> Looks good, just had a few more comments and it looks like there are some lint failures.

Thanks Pritam. Fixed all the cmts, PTAL

bowangbj committed:
lint on "More sharded_tensor creation ops: harded_tensor.zeros, sharded_tensor.full, sharded_tensor.rand"


Summary:

Test Plan:

$ python test/distributed/_sharded_tensor/test_sharded_tensor.py  --v

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D30472621](https://our.internmc.facebook.com/intern/diff/D30472621)

[ghstack-poisoned]

bowangbj commented:
@bowangbj has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30472621).

bowangbj commented:
The only tricky lint error is 

./torch/distributed/_sharded_tensor/__init__.py:1:1: C101 Coding magic comment not found
Please fix the above Flake8 warnings.
Error: Process completed with exit code 1.

I run $flake8 __init__.py locally which produced no error.

bowangbj committed:
lint C101 on "More sharded_tensor creation ops: harded_tensor.zeros, sharded_tensor.full, sharded_tensor.rand"


Summary:

Test Plan:

$ python test/distributed/_sharded_tensor/test_sharded_tensor.py  --v

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D30472621](https://our.internmc.facebook.com/intern/diff/D30472621)

[ghstack-poisoned]

bowangbj committed:
lint C101 on "More sharded_tensor creation ops: harded_tensor.zeros, sharded_tensor.full, sharded_tensor.rand"


Summary:

Test Plan:

$ python test/distributed/_sharded_tensor/test_sharded_tensor.py  --v

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D30472621](https://our.internmc.facebook.com/intern/diff/D30472621)

[ghstack-poisoned]

bowangbj committed:
add test_full_with_dtype_overridden on "More sharded_tensor creation ops: harded_tensor.zeros, sharded_tensor.full, sharded_tensor.rand"


Summary:

Test Plan:

$ python test/distributed/_sharded_tensor/test_sharded_tensor.py  --v

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D30472621](https://our.internmc.facebook.com/intern/diff/D30472621)

[ghstack-poisoned]
